File: sdk/mcp/sdk.go
package mcp

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"sync"
	"sync/atomic"
	"time"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type Server struct {
	name        string
	version     string
	description string
	platformURL string
	token       string
	serverID    string
	mcpServer   *server.MCPServer
	httpServer  *server.StreamableHTTPServer

	requestsTotal   int64
	requestsSuccess int64
	requestsError   int64
	latencySum      float64
	latencyMu       sync.Mutex
	toolCalls       map[string]int64
	toolMu          sync.RWMutex
}

type Config struct {
	Name        string
	Version     string
	Description string
}

func NewServer(config Config) *Server {
	platformURL := os.Getenv("MANAGING_UP_PLATFORM_URL")
	if platformURL == "" {
		platformURL = "http://localhost:8080"
	}

	name := config.Name
	if name == "" {
		name = os.Getenv("MANAGING_UP_NAME")
	}
	version := config.Version
	if version == "" {
		version = os.Getenv("MANAGING_UP_VERSION")
		if version == "" {
			version = "1.0.0"
		}
	}

	return &Server{
		name:        name,
		version:     version,
		description: config.Description,
		platformURL: platformURL,
		token:       os.Getenv("MANAGING_UP_TOKEN"),
		mcpServer: server.NewMCPServer(name, version,
			server.WithToolCapabilities(true),
			server.WithLogging(),
		),
		toolCalls: make(map[string]int64),
	}
}

func (s *Server) AddTool(tool mcp.Tool, handler func(ctx context.Context, request mcp.CallToolRequest) (*mcp.CallToolResult, error)) {
	s.toolMu.Lock()
	defer s.toolMu.Unlock()
	s.toolCalls[tool.Name] = 0
	s.mcpServer.AddTool(tool, handler)
}

type Metrics struct {
	RequestsTotal   int64            `json:"requests_total"`
	RequestsSuccess int64            `json:"requests_success"`
	RequestsError   int64            `json:"requests_error"`
	LatencySeconds  float64          `json:"latency_seconds"`
	ToolCalls       map[string]int64 `json:"tool_calls"`
}

func (s *Server) Metrics() Metrics {
	s.toolMu.RLock()
	defer s.toolMu.RUnlock()
	s.latencyMu.Lock()
	latency := s.latencySum
	s.latencyMu.Unlock()
	return Metrics{
		RequestsTotal:   atomic.LoadInt64(&s.requestsTotal),
		RequestsSuccess: atomic.LoadInt64(&s.requestsSuccess),
		RequestsError:   atomic.LoadInt64(&s.requestsError),
		LatencySeconds:  latency,
		ToolCalls:       s.toolCalls,
	}
}

type RegistrationRequest struct {
	Name          string `json:"name"`
	Version       string `json:"version"`
	Description   string `json:"description,omitempty"`
	URL           string `json:"url,omitempty"`
	TransportType string `json:"transport_type"`
	Status        string `json:"status"`
}

type RegistrationResponse struct {
	ID     string `json:"id"`
	Name   string `json:"name"`
	Status string `json:"status"`
}

func (s *Server) Register(ctx context.Context) (*RegistrationResponse, error) {
	req := RegistrationRequest{
		Name:          s.name,
		Version:       s.version,
		Description:   s.description,
		TransportType: "http",
		Status:        "pending",
	}

	body, err := json.Marshal(req)
	if err != nil {
		return nil, fmt.Errorf("failed to marshal request: %w", err)
	}

	url := s.platformURL + "/api/v1/mcp-servers"
	httpReq, err := http.NewRequestWithContext(ctx, http.MethodPost, url, bytes.NewReader(body))
	if err != nil {
		return nil, err
	}

	httpReq.Header.Set("Content-Type", "application/json")
	if s.token != "" {
		httpReq.Header.Set("Authorization", "Bearer "+s.token)
	}

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(httpReq)
	if err != nil {
		return nil, fmt.Errorf("failed to register: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusCreated {
		return nil, fmt.Errorf("registration failed with status: %d", resp.StatusCode)
	}

	var result struct {
		Data RegistrationResponse `json:"data"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, fmt.Errorf("failed to decode response: %w", err)
	}

	s.serverID = result.Data.ID
	return &result.Data, nil
}

func (s *Server) StartStdio(ctx context.Context) error {
	if err := server.ServeStdio(s.mcpServer); err != nil {
		return fmt.Errorf("failed to serve stdio: %w", err)
	}
	return nil
}

func (s *Server) StartHTTP(ctx context.Context, addr string) error {
	s.httpServer = server.NewStreamableHTTPServer(s.mcpServer)
	log.Printf("Starting MCP server %s@%s on %s", s.name, s.version, addr)
	log.Printf("MCP endpoint: http://localhost%s/mcp", addr)
	return s.httpServer.Start(addr)
}

func (s *Server) recordRequest(success bool, latency time.Duration) {
	atomic.AddInt64(&s.requestsTotal, 1)
	if success {
		atomic.AddInt64(&s.requestsSuccess, 1)
	} else {
		atomic.AddInt64(&s.requestsError, 1)
	}
	s.latencyMu.Lock()
	s.latencySum += latency.Seconds()
	s.latencyMu.Unlock()
}

func (s *Server) recordToolCall(toolName string) {
	s.toolMu.Lock()
	defer s.toolMu.Unlock()
	s.toolCalls[toolName]++
}

File: sdk/typescript/src/mcp/sdk.ts
/**
 * Managing-Up MCP SDK for TypeScript/Node.js
 */

export enum TransportType {
  STDIO = "stdio",
  HTTP = "http",
  SSE = "sse",
}

export interface MCPServerConfig {
  name: string;
  version: string;
  description?: string;
  transportType: TransportType;
  url?: string;
  command?: string;
  args?: string[];
  env?: Record<string, string>;
}

export interface Metrics {
  requests_total: number;
  requests_success: number;
  requests_error: number;
  latency_seconds: number;
  tool_calls: Record<string, number>;
}

export interface ToolHandler {
  (args: Record<string, unknown>): Promise<string>;
}

export interface ToolDefinition {
  name: string;
  description: string;
  inputSchema: Record<string, unknown>;
  handler: ToolHandler;
}

export class MCPServer {
  private config: MCPServerConfig;
  private platformUrl: string;
  private token: string;
  private serverId?: string;
  private tools: Map<string, ToolDefinition> = new Map();
  private metrics: Metrics = {
    requests_total: 0,
    requests_success: 0,
    requests_error: 0,
    latency_seconds: 0,
    tool_calls: {},
  };

  constructor(config: MCPServerConfig) {
    this.config = config;
    this.platformUrl = process.env.MANAGING_UP_PLATFORM_URL || "http://localhost:8080";
    this.token = process.env.MANAGING_UP_TOKEN || "";
  }

  static fromEnv(): MCPServer {
    return new MCPServer({
      name: process.env.MANAGING_UP_NAME || "typescript-mcp-server",
      version: process.env.MANAGING_UP_VERSION || "1.0.0",
      description: process.env.MANAGING_UP_DESCRIPTION,
      transportType: (process.env.MANAGING_UP_TRANSPORT_TYPE as TransportType) || TransportType.HTTP,
      url: process.env.MANAGING_UP_URL,
    });
  }

  addTool(definition: ToolDefinition): void {
    this.tools.set(definition.name, definition);
  }

  async register(): Promise<{ id?: string; name?: string; status?: string } | null> {
    const url = `${this.platformUrl}/api/v1/mcp-servers`;
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
    };
    if (this.token) {
      headers["Authorization"] = `Bearer ${this.token}`;
    }

    const payload = {
      name: this.config.name,
      version: this.config.version,
      description: this.config.description || "",
      transport_type: this.config.transportType,
      url: this.config.url || "",
      status: "pending",
    };

    try {
      const response = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify(payload),
      });

      if (response.status === 201) {
        const data = await response.json();
        this.serverId = data.data?.id;
        console.log(`Registered with Managing-Up platform: ${this.serverId}`);
        return data.data;
      } else {
        console.warn(`Registration failed with status ${response.status}`);
        return null;
      }
    } catch (error) {
      console.warn(`Registration failed: ${error}`);
      return null;
    }
  }

  recordRequest(success: boolean, latencyMs: number): void {
    this.metrics.requests_total++;
    if (success) {
      this.metrics.requests_success++;
    } else {
      this.metrics.requests_error++;
    }
    this.metrics.latency_seconds += latencyMs / 1000;
  }

  recordToolCall(toolName: string): void {
    if (!this.metrics.tool_calls[toolName]) {
      this.metrics.tool_calls[toolName] = 0;
    }
    this.metrics.tool_calls[toolName]++;
  }

  getMetrics(): Metrics {
    return { ...this.metrics };
  }

  async handleRequest(request: {
    method: string;
    params?: { name?: string; arguments?: Record<string, unknown> };
    id?: string | number;
  }): Promise<{ jsonrpc: string; id?: string | number; result?: unknown; error?: { code: number; message: string } }> {
    const startTime = Date.now();

    try {
      if (request.method === "tools/list") {
        const tools = Array.from(this.tools.values()).map((t) => ({
          name: t.name,
          description: t.description,
          inputSchema: t.inputSchema,
        }));
        return { jsonrpc: "2.0", id: request.id, result: { tools } };
      }

      if (request.method === "tools/call") {
        const toolName = request.params?.name;
        const arguments_ = request.params?.arguments || {};

        if (toolName && this.tools.has(toolName)) {
          const tool = this.tools.get(toolName)!;
          const result = await tool.handler(arguments_);
          this.recordToolCall(toolName);
          this.recordRequest(true, Date.now() - startTime);
          return {
            jsonrpc: "2.0",
            id: request.id,
            result: { content: [{ type: "text", text: String(result) }] },
          };
        } else {
          this.recordRequest(false, Date.now() - startTime);
          return {
            jsonrpc: "2.0",
            id: request.id,
            error: { code: -32601, message: `Tool not found: ${toolName}` },
          };
        }
      }

      return {
        jsonrpc: "2.0",
        id: request.id,
        error: { code: -32601, message: "Method not found" },
      };
    } catch (error) {
      this.recordRequest(false, Date.now() - startTime);
      return {
        jsonrpc: "2.0",
        id: request.id,
        error: { code: -32603, message: String(error) },
      };
    }
  }

  async serveStdio(): Promise<void> {
    const readline = await import("readline");
    const rl = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      terminal: false,
    });

    for await (const line of rl) {
      try {
        const request = JSON.parse(line.trim());
        const response = await this.handleRequest(request);
        if (response) {
          console.log(JSON.stringify(response));
        }
      } catch {
        // Ignore parse errors
      }
    }
  }

  createHandler(): (request: Request) => Promise<Response> {
    return async (request: Request): Promise<Response> => {
      if (request.method === "GET") {
        return new Response(JSON.stringify({ tools: Array.from(this.tools.keys()) }), {
          headers: { "Content-Type": "application/json" },
        });
      }

      try {
        const body = await request.json();
        const response = await this.handleRequest(body);
        return new Response(JSON.stringify(response), {
          headers: { "Content-Type": "application/json" },
        });
      } catch (error) {
        return new Response(
          JSON.stringify({ jsonrpc: "2.0", error: { code: -32700, message: "Parse error" } }),
          { status: 400, headers: { "Content-Type": "application/json" } }
        );
      }
    };
  }
}

export default MCPServer;
